src/common/hist_util.cu

```diff
@@ -324,9 +316,27 @@ struct GPUSketcher {
 
       // add cuts into sketches
       thrust::copy(cuts_d_.begin(), cuts_d_.end(), cuts_h_.begin());
+#pragma omp parallel for schedule(static)
       for (int icol = 0; icol < num_cols_; ++icol) {
-        summaries_[icol].MakeFromSorted(&cuts_h_[n_cuts_ * icol], n_cuts_cur_[icol]);
-        sketches_[icol].PushSummary(summaries_[icol]);
+        WXQSketch::SummaryContainer summary;
+        summary.Reserve(n_cuts_);
+        summary.MakeFromSorted(&cuts_h_[n_cuts_ * icol], n_cuts_cur_[icol]);
+
+        // Spin in user land until we we can grab this lock
+        struct ColLocker {
+         public:
+          explicit ColLocker(std::atomic_flag *lock)
+            : col_lock_(lock) {
+              while (col_lock_->test_and_set()) {}
+            }
+          ~ColLocker() {
+            col_lock_->clear();
+          }
+         private:
+          std::atomic_flag *col_lock_;
+        }lock(sketch_container_->col_locks_[icol].get());
```

Comment:
Hmm I guess I don't quite understand this code. Since you are looping through the columns, each thread is working on a different block of columns, why do you still need to lock on the column?

Reply:
rows in a batch are split across multiple devices and the rows across those devices can be aggregating sketches for the same column.

Reply:
Oh right, this is in a device shard. Since each shard needs to process all the columns, wouldn't this introduce a lot of contention?

Reply:
sure - if all the shards tries to push the summary for the same column at the same time. from whatever profiling i have done on xgboost for some of the datasets i tried, the hotspots in this method seem to be in computing the column cuts and presizing the quantile sketch summary. since we are spinning here, it will burn some cpu if they are contending as opposed to a more expensive lock acquisition via a mutex.